src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -190,135 +238,162 @@ public long ramBytesUsed() {
+            return;
+        }
+        for (Map.Entry<String, Map<FsChannelReader, RuntimeException>> file : ChannelReference.openedFiles.entrySet()) {
+            IOUtils.closeWhileHandlingException(file.getValue().keySet());
+        }
+    }
+
+
+    /** gets a list of unreferenced files (only works if assertions are enabled, returns an empty array o.w.) */
+    public String[] getUnreferenced() throws IOException {
+        if (ChannelReference.openedFiles == null) {
+            return Strings.EMPTY_ARRAY; // not supported
+        }
+        ArrayList<String> result = new ArrayList<>();
+        try (ReleasableLock _ = writeLock.acquire()) {
             for (Path location : locations) {
                 try (DirectoryStream<Path> stream = Files.newDirectoryStream(location, TRANSLOG_FILE_PREFIX + "[0-9]*")) {
                     for (Path file : stream) {
-                        if (isReferencedTranslogFile(file) == false) {
-                            try {
-                                logger.trace("delete unreferenced translog file: " + file);
-                                Files.delete(file);
-                                deleted++;
-                            } catch (Exception ex) {
-                                logger.debug("failed to delete " + file, ex);
-                            }
+                        final long id = parseIdFromFileName(file);
+                        if (id >= 0 && ChannelReference.openedFiles.containsKey(file.toString()) == false) {
+                            result.add(file.toString());
                         }
                     }
                 }
             }
-        } catch (IOException ex) {
-            logger.debug("failed to clear unreferenced files ", ex);
-        } finally {
-            rwl.writeLock().unlock();
         }
-        return deleted;
+        return result.toArray(Strings.EMPTY_ARRAY);
     }
 
     @Override
-    public void newTranslog(long id) throws TranslogException, IOException {
-        rwl.writeLock().lock();
-        try {
-            FsTranslogFile newFile;
-            long size = Long.MAX_VALUE;
-            Path location = null;
-            for (Path file : locations) {
-                long currentFree = Files.getFileStore(file).getUsableSpace();
-                if (currentFree < size) {
-                    size = currentFree;
-                    location = file;
-                } else if (currentFree == size && ThreadLocalRandom.current().nextBoolean()) {
-                    location = file;
+    public void committedTranslogId(final long translogId) throws FileNotFoundException {
+        try (ReleasableLock _ = writeLock.acquire()) {
+            logger.trace("updating translogs on commit of [{}]", translogId);
+            if (translogId < lastCommitedTranslogId) {
+                throw new ElasticsearchIllegalArgumentException("committed translog id can only go up (current ["
```

Comment:
Can do though I wonder we ask? this is a parameter which is inconsistent with current state. EIAE seems better?